src/qibo/callbacks.py

```diff
@@ -330,6 +341,7 @@ def apply(self, backend, state):
                 "The Hamiltonian is degenerate. Using eigenvalue {} "
                 "to calculate gap.".format(excited)
             )
+        self.append(gap)
```

Comment:
This is correct, but if you check a few lines above this callback has additional returns before that, which may be triggered in some cases and if they do, it will not append. I think the safest in this case would be to append before every `return`. This should also fix the failing tests.